pkg/api/message/service.go

```diff
@@ -117,19 +122,104 @@ func (s *Service) SubscribeEnvelopes(
 	}
 }
 
+// Pulls from DB and sends to client, updating the query's last seen cursor, until
+// the stream has caught up to the latest in the database.
+func (s *Service) catchUpFromCursor(
+	stream message_api.ReplicationApi_SubscribeEnvelopesServer,
+	query *message_api.EnvelopesQuery,
+	logger *zap.Logger,
+) error {
+	log := logger.With(zap.String("stage", "catchUpFromCursor"))
+	if query.GetLastSeen() == nil {
+		log.Debug("Skipping catch up")
+		// Requester only wants new envelopes
+		return nil
+	}
+
+	cursor := query.LastSeen.GetNodeIdToSequenceId()
+	// GRPC does not distinguish between empty map and nil
+	if cursor == nil {
+		cursor = make(map[uint32]uint64)
+		query.LastSeen.NodeIdToSequenceId = cursor
+	}
+
+	log.Debug("Catching up from cursor", zap.Any("cursor", cursor))
+	for {
+		rows, err := s.fetchEnvelopes(stream.Context(), query, maxRequestedRows)
+		log.Debug("Fetched envelopes", zap.Any("rows", rows))
+		if err != nil {
+			return err
+		}
+		envs := make([]*envelopes.OriginatorEnvelope, 0, len(rows))
+		for _, r := range rows {
+			env, err := envelopes.NewOriginatorEnvelopeFromBytes(r.OriginatorEnvelope)
+			if err != nil {
+				// We expect to have already validated the envelope when it was inserted
+				logger.Error("could not unmarshal originator envelope", zap.Error(err))
+				continue
+			}
+			envs = append(envs, env)
+		}
+		err = s.sendEnvelopes(stream, query, envs)
+		if err != nil {
+			return status.Errorf(codes.Internal, "error sending envelopes: %v", err)
+		}
+		if len(rows) < int(maxRequestedRows) {
+			// There were no more envelopes in DB at time of fetch
+			break
+		}
+		// TODO(rich): Determine default rate limits and set query interval to match
+		time.Sleep(50 * time.Millisecond)
```

Comment:
Would love for us to work out what rate limits we want to impose by default (e.g. if a client sends many requests), and what setting to use here that would fall within that rate limit (which is the equivalent of requesting every 50ms)

Reply:
The actual bookkeeping of the rate limits is going to be interesting here, since a query that involves catchup could yield 1 database query or >10,000.

I wonder if we are going to have to meter based on the actual number of underlying database queries vs some forward projection of how many queries the request is going to take. Like we would decrement your rate limit on every page of results pulled.

What about backwards from a target of 10 requests/second as a generous but not extravagant starting place?

Reply:
Sounds good - updated to a 100ms paging interval for now with a 10 req/sec target in mind. We may need to adjust this on a more granular level for firehose-type applications, like node<>node replication. Will follow up on that when we are reading the JWT and poking a hole in the default rate limit